Linux: stop flickering and stretching on resize

adlr opened the pull request:
A few commits to address the concerns of flickering and content stretching on https://github.com/ghostty-org/ghostty/issues/2446

First, tell OpenGL the true size it has at the time drawFrame is called, so that it's not using an older size and thus stretching content.

Second, revert a patch that freed up memory from unused cell capacity in the OpenGL renderer, and replace it with a new patch that frees up the memory. The problem with the earlier patch was it was freeing the cells fully, then OpenGL would display content (which was blank!) and then rendering would complete and cells would be set again. This was leading to occasional blank frames during render. The new change keeps memory down, as was the goal with the earlier patch, but keeps the old cells around a bit longer in case they are needed to paint.

This is my first change on this codebase and in the Zig language; any advice is welcome!

NOTE: I am hoping CI will test that this doesn't break macOS. I don't have a mac handy to test.

adlr committed:
OpenGL: Tell GL the true screen size so it can avoid stretching content

adlr committed:
Revert "renderer/opengl: reclaim cell buffer memory when resizing"

This was causing flicker during resize

This reverts commit 06edc311a1e3fa4611f76fe47bd6e7580257bbc3.

adlr committed:
Free up cells memory on resize

mitchellh reviewed (commented):
Thanks for taking this on, I appreciate it. Let me take a closer look and see if any of what you did makes sense anywhere else, but the fundamentals of it are sound!

jcollie commented:
I tried it out locally and it definitely seemed to help. Most notably when I had helix open to a blank scratch buffer. I can't really comment on the code as I have little experience with the OpenGL code.

mitchellh commented:
Thank you! I modified and made a new PR #7155. I gave you coauthor credit. The real change I made was around `setTrueSize`. Instead of that, I changed it to query the viewport dimensions on every frame draw. GTK guarantees the GL context viewport is updated by the time draw frame is called.

adlr commented:
> Thank you! I modified and made a new PR #7155. I gave you coauthor credit. The real change I made was around `setTrueSize`. Instead of that, I changed it to query the viewport dimensions on every frame draw. GTK guarantees the GL context viewport is updated by the time draw frame is called.

Thanks so much! I didn't know that was possible and it looks like a cleaner solution. Really appreciate the quick response here and I'm glad this is in.
